### ISSUE TITLE
DomUs are not stopped before dom0 shutdown

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4368]
### Qubes OS version:

`Qubes release 4.0 (R4.0)`

### Affected component(s): 

qubes-core-admin

---

### Steps to reproduce the behavior:

Shut down Qubes from the UI

### Expected behavior:

Qubes shuts down normally

### Actual behavior:

Qubes hangs for a while waiting for block devices to stop, and only shuts down when the corresponding units time out.

### General notes:

Shutting down all domUs before shutting down dom0 fixes this. 

---

**Comment by andrewdavidwong** [comment]
Duplicate of #1581

**Comment by andrewdavidwong** [comment]
This appears to be a [duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) of an existing issue. If you believe this is not really a duplicate, please leave a comment briefly explaining why. We'll be happy to take another look and, if appropriate, reopen this issue. Thank you.

**Comment by DemiMarie** [comment]
@andrewdavidwong I don’t think this is a duplicate.  #1581 was due to calling X functions from a signal handler, which isn’t allowed.  Furthermore, this problem is reproduceable with a fully updated dom0.  The symptoms are that Qubes stalls for 90s waiting for block device shutdown to timeout.

**Comment by constantoverride** [comment]
>  Qubes stalls for 90s waiting for block device shutdown to timeout.

I guess the reason you would hit this now(assuming you're up to date) is because the shutdown of VMs and the "shutdown" of block devices are both running in parallel, so if your VMs take more than 90s to shutdown, you'll get that timeout error from block devices.

See how long does the following `preshutdown` script (aka shutting down all running VMs script) take to finish, while you're in X:
```bash
#!/bin/bash
  
tput -Txterm-256color setab 2;echo "Starting $0 `date` `cat /proc/uptime`";echo;tput -Txterm-256color sgr0;
xl list
time qvm-shutdown --verbose --all --wait; ec="$?"
echo "exitcode: '$ec'"
time while xl list|grep -q -F '(null)'; do xl list;sleep 1; done #this only happens if there are any qubes-guid pids active (eg. pkill -9 qubes-guid is required to never see any of these '(null)'s                                                                                                                                  
tput -Txterm-256color setab 1;echo "Ending $0 on `date` `cat /proc/uptime`";echo;tput -Txterm-256color sgr0;
exit $ec
```
(run as `./preshutdown`)
If it's like less than 80 seconds, then are you sure you're up to date? 
ie. 
```
[ctor@dom0 ~]$ rpm -qi qubes-gui-dom0-4.0.9-1.fc25
Name        : qubes-gui-dom0
Version     : 4.0.9
Release     : 1.fc25
Architecture: x86_64
Install Date: Sun 16 Sep 2018 12:01:06 AM CEST
Group       : Qubes
Size        : 167095
License     : GPL
Signature   : RSA/SHA256, Sat 15 Sep 2018 11:30:42 PM CEST, Key ID 1848792f9e2795e9
Source RPM  : qubes-gui-dom0-4.0.9-1.fc25.src.rpm
Build Date  : Sat 15 Sep 2018 11:30:34 PM CEST
Build Host  : build-fedora4
Relocations : (not relocatable)
Vendor      : Invisible Things Lab
URL         : http://www.qubes-os.org
Summary     : The Qubes GUI virtualization (Dom0 side)
Description :
The Qubes GUI virtualization infrastructure that needs to be installed in Dom0.
```

You can see that there's `TimeoutStopSec=180` in `/usr/lib/systemd/system/qubes-core.service` which is a file owned by `qubes-core-dom0-4.0.29-1.fc25.x86_64`
```
[ctor@dom0 ~]$ rpm -qf /usr/lib/systemd/system/qubes-core.service
qubes-core-dom0-4.0.29-1.fc25.x86_64
```
However, that `180` doesn't really matter in your case because the default timeout of 90 sec will be hit by whatever is trying to shutdown the block devices well before this 180sec one is done. Assuming again, that shutting down the VMs takes that long for you (for me it's never above 30 sec).

Something systemd-related causes both types of shutdowns to occur concurrently. Technically, before #1581 was fixed, X was preventing `qubes-core.service` from reaching its Stop stage, but the block devices shutdown was already started! So now that 1581 is fixed, X isn't preventing, so `qubes-core.service`'s Stop actions are running (in parallel) with the block devices shutdown, which will definitely timeout after 90sec if stopping all of your running VMs takes more than 90sec. In other words, `qubes-core.service` is expected to finish stopping and thus not keeping those block devices in use anymore, well before the stopping of those block devices times out (90sec).

So, you could try, I guess, increasing the default timeout, in file `/etc/systemd/system.conf`:
```
#DefaultTimeoutStartSec=90s
#DefaultTimeoutStopSec=90s
```
I guess you only need to change `DefaultTimeoutStopSec`.

A proper fix would be making sure those block devices aren't being stopped until after `qubes-core.service` is fully stopped(eg. all VMs are shutdown), but good luck doing that with systemd, imho :D
